src/thermal_conductivity/scattering.f90

```diff
@@ -0,0 +1,347 @@
+                velnorm = norm2(dr%iq(sr%q1(il))%vel(:, sr%b1(il)))
+                if (velnorm .gt. lo_phonongroupveltol) then
+                    buf = buf + velnorm/opts%mfp_max
+                end if
+            end if
+            ! Now we can update the linewidth for this mode
+            buf_lw(sr%q1(il), sr%b1(il)) = buf
+
+            if (mw%talk .and. lo_trueNtimes(il, 127, sr%nlocal_point)) then
+                call lo_progressbar(' ... computing scattering amplitude', il, sr%nlocal_point, walltime() - t0)
+            end if
+        end do
+        if (mw%talk) call lo_progressbar(' ... computing scattering amplitude', sr%nlocal_point, sr%nlocal_point, walltime() - t0)
+
+        ! Reduce the linewidth
+        call mw%allreduce('sum', buf_lw)
+
+        ! Distribute it after fixing the degeneracies
+        do q1 = 1, qp%n_irr_point
+            do b1 = 1, dr%n_mode
+                if (dr%iq(q1)%omega(b1) .lt. lo_freqtol) cycle
+                ! First we fix the degeneracy
+                f0 = 0.0_r8
+                do j = 1, dr%iq(q1)%degeneracy(b1)
+                    b2 = dr%iq(q1)%degenmode(j, b1)
+                    f0 = f0 + buf_lw(q1, b2)
+                end do
+                f0 = f0/real(dr%iq(q1)%degeneracy(b1), r8)
+                do j = 1, dr%iq(q1)%degeneracy(b1)
+                    b2 = dr%iq(q1)%degenmode(j, b1)
+                    buf_lw(q1, b2) = f0
+                end do
+                ! Now we can set the linewidth
+                dr%iq(q1)%linewidth(b1) = buf_lw(q1, b1)
+
+                ! While we are at it, we can set other things
+                dr%iq(q1)%qs(b1) = 2.0_r8*dr%iq(q1)%linewidth(b1)
+                velnorm = norm2(dr%iq(q1)%vel(:, b1))
+                if (velnorm .gt. lo_phonongroupveltol) then
+                    dr%iq(q1)%mfp(:, b1) = dr%iq(q1)%vel(:, b1)/dr%iq(q1)%qs(b1)
+                    dr%iq(q1)%scalar_mfp(b1) = velnorm/dr%iq(q1)%qs(b1)
+                    dr%iq(q1)%F0(:, b1) = dr%iq(q1)%mfp(:, b1)
+                    dr%iq(q1)%Fn(:, b1) = dr%iq(q1)%F0(:, b1)
+                end if
+            end do
+        end do
+        call mem%deallocate(buf_lw, persistent=.false., scalable=.false., file=__FILE__, line=__LINE__)
+    end block scatt
+
+    symmetrize_scatmat: block
+        !> To hold the q-point in reduced coordinates
+        real(r8), dimension(3) :: qv2, qv2p
+        !> To get the index of the new triplet on the fft_grid
+        integer, dimension(3) :: gi
+        !> Some buffer
+        real(r8), dimension(:), allocatable :: buf
+        !> Integer for do loops and so on
+        integer :: il, jl, q1, j, k, q2, b2, q2p, n
+
+        call tmr%tick()
+        if (mw%talk) write (*, *) '... symmetrizing scattering matrix'
+
+        ! We use the relation Xi_{R*q, R*q'} = Xi_{q, q'''} to enforce the symmetry of Xi
+        ! TODO actually use this symmetry to reduce the number of scattering to compute
```

Comment:
We are using symmetries to obtain irreducible triplets/quartets for the scattering, but not the irreducible pair for the scattering matrix. For this, we only average over equivalent points.
I'm not entirely sure that we can reduce the number of scattering to compute with the irreducible pair, but it might be worth having a look.